### ISSUE TITLE
Investigate `kafka` -> `remap` -> `elasticsearch`/`aws_s3`

### ISSUE THREAD

**Issue by tobz** [vectordotdev/vector#8636]
This issue is a record of and parent to the investigation of vector's performance when configured like so:

```
[sources.internal_metrics]
type = "internal_metrics"

[sources.kafka_gen]
type = "kafka"
bootstrap_servers = "kafka:9092"
group_id = "kafka_json_es_s3"
topics = ["kafka_json_es_s3"]

[transforms.msg_split]
type = "remap"
inputs = ["kafka_gen"]
source = '''
. = split(string!(.message), "\n")
'''

[sinks.prometheus]
type = "prometheus_exporter"
inputs = ["internal_metrics"]
address = "0.0.0.0:9598"

[sinks.aws_es]
type = "elasticsearch"
inputs = ["msg_split"]
endpoint = "https://elasticsearch"

[sinks.aws_s3]
type = "aws_s3"
inputs = ["msg_split"]
bucket = "results-bucket"
region = "us-east-2"
key_prefix = "kafka_json_es_s3/date=%F"
encoding.codec = "ndjson"
batch.max_bytes = 52428800
batch.timeout_secs = 3```

**Comment by tobz** [comment]
Running the soak test as configured above, with the latest nightly, current results in a runaway memory leak that kills the process after a short period of time: 5-10 minutes at most.

**Comment by tobz** [comment]
After some head banging, it seems like the memory leak is related to a runaway number of request retries.

If the downstream sink returns that it's overwhelmed, or responds in a way that would trigger a retry, we'll dutifully attempt that retry.  As it turns out, our Tower service stack is configured in the following order: rate limit, retries, then ARC.  Thus, we don't actually subject retries to rate limiting, which means we're still letting normal events through, which compete with the retries.  This situation almost completely ensures that under steady state load, Vector can't possibly clear the backlog of retries unless the downstream service becomes able to handle the load.

We should be able to swap the retry layer with the rate limit layer to ensure that we're rate limiting retries in the same fashion as normal requests.

**Comment by tobz** [comment]
Opened https://github.com/timberio/vector/issues/8647 to track the above issue.

**Comment by tobz** [comment]
Getting the Elasticsearch sink to behave well (not overload the downstream service) while also maximizing performance has been a bit of a bear.  I tried to go back to basics here to analyze the performance/reliability as we tweak knobs.  These tests are based on a branch of Vector that addresses the aforementioned ordering issue of retries vs rate limiting. 
 As well, we removed the S3 sink to fully isolate the `Kafka -> remap -> ES` path.  Our Kafka data generation is right around 65MB/second.

Our first run is targeting `http-sinkhole`, which is a daemon that listens for HTTP requests and always responds back with a 200 OK.  Crucially, though, it waits 15ms before responding to simulate some downstream load.  We're using all of the default batching and request limits in Vector.  This variant is `unbound_sinkhole`.

This variant performs well, as expected.  We observe no unbounded memory growth and performance is flat/consistent.  It easily keeps up with the incoming data volume from Kafka.  Here's a screenshot of the overall throughput of the Kafka data generator, and the Vector `kafka` and `elasticsearch` components:

![Screen Shot 2021-08-11 at 10 11 46 AM](https://user-images.githubusercontent.com/222615/129044841-f7324afc-1d39-4299-9e0a-701fa792189c.png)

The peak before all of the lines converge is actually the Kafka consumer being behind the current topic offset, so it races to catch up, and then is consuming essentially in lockstep with the data generation.

We use this as our baseline: we know the ES sink can shuffle off the data as fast as we consume it assuming the downstream is healthy and responding.

**Comment by tobz** [comment]
Our next run is still targeting `http-sinkhole`, but we've adjusted `request.rate_limit_num` to a value of `1`.  This is simply for us to demonstrate that the ES sink is batching by default, and assert those batching values.  As expected, the default batch size is in fact 10MB, with a 1 second timeout.  This variant is `1rps_sinkhole`.

Not much else to say here besides things look exactly like we'd expect:

![Screen Shot 2021-08-11 at 10 22 43 AM](https://user-images.githubusercontent.com/222615/129046736-0d965a7a-3af7-4cf2-9db7-56995770e940.png)

**Comment by tobz** [comment]
Our next run is now targeting our AWS Elasticsearch cluster.  This cluster is compromised of 3x dedicated master nodes -- `c6g.large.elasticsearch`, with 3x data nodes of the same instance type, with 512GB EBS volumes (gp2 SSDs).  In practice, this is indeed a fairly small-ish cluster, although our goal is to maximize our throughput to it while not overloading it or having Vector run into reliability issues.

We've reset all of the configuration for the ES sink: default request limits and batching.  This variant is `default_aws_es`.

This variant does manage to sling a lot of data at ES, but with a flurry of retries that lead to Vector running into an OOM situation a few minutes after it launches:

![Screen Shot 2021-08-11 at 10 56 03 AM](https://user-images.githubusercontent.com/222615/129052717-37cff8cb-c2d4-444f-bf50-e4a625cb75db.png)

The main problem here is that AWS ES will respond back quickly when its write queue is full, which triggers Vector's retry logic.  The request is retried after waiting for 1 second.  As the default `request.rate_limit_num` is 2^63, and the default request concurrency is `1024`, Vector is only limited by the batching behavior and concurrency limits in terms of generating more requests, beyond the initial request for a particular event.  This leads to a stacking effect where we continue generating more and more requests that are hitting the downstream ES cluster.  Each of those requests also has to be cloned to be re-sent, and at 10MB batches, those stacked requests quickly chew up system memory.

**Comment by tobz** [comment]
Our next run is similar to `concurrency_5_aws_es`, but instead of a fixed concurrency of 5 requests, we've toggled on adaptive concurrency, or ARC. ARC observes the performance of responses from the downstream service, and adjusts concurrency limits on-the-fly to try and avoid overloading those downstream services.  This variant is `concurrency_arc_aws_es`.

Results using ARC over a fixed concurrency of 5 requests don't seem to be meaningfully distinguishable in terms of performance, although throughput is a bit more flat/well paced:

![Screen Shot 2021-08-11 at 1 22 06 PM](https://user-images.githubusercontent.com/222615/129074740-bd7e82a0-0cac-43cc-8009-3840084bb183.png)

**Comment by tobz** [comment]
We've made a few changes here, and we're correlating a few things from multiple runs, but here it goes....

Overall, our goal for Vector is not only to be performant at shipping observability data, but to do so _reliably_.  Part of that is introducing features like Adaptive Concurrency (ARC), which sense downstream system load/latency through measuring the round trip time of requests, and adjusting concurrency until our requests are not perturbing the performance of the system above a certain threshold.

In practice, what this means is that undersized downstream systems may not be fully maxed out in terms of throughout.  While a brute force approach might yield measurably better results, we err on the side of keeping latencies stable, as a proxy for system responsiveness.

All of our tests so far have been with a small Elasticsearch cluster: 3 master nodes, 3 data nodes, all of which were `c6g.large.elasticsearch` machines, so 2 vCPUs and 8GB memory, and 512GB gp2 SSD volumes..  Elasticsearch is a Java application which is stereotypically memory-intensive, but I still wanted to see what Vector could do with that small configuration.  The results above bore out that experiment, and it wasn't great.

We've upsized the Elasticsearch cluster to `i3.2xlarge.elasticsearch` for the data nodes, which are 8 vCPUs and 64GB memory, with 1.9TB of local NVMe storage.  In short, much faster.  We've also made sure that the shard configuration is optimized to spread out load across the entire cluster of data nodes, with a tweaked `refresh_interval` of 30 seconds to ensure that our high rate of ingestion isn't constantly battling index refreshing.

With those changes in place, the `default_aws_es` variant can run without any sort of memory leak-esque behavior, and keeps pace with the Kafka ingestion rate.  As well, using @bruceg's ARC branch, which actually optimized some conservative defaults that previously limited the maximum performance when using ARC, we can similarly keep up with the Kafka ingestion rate.  Below is our throughput graph running the `concurrency_arc_aws_es` variant:

![Screen Shot 2021-08-12 at 7 04 20 PM](https://user-images.githubusercontent.com/222615/129280956-93ea5ab9-6d13-47ec-9e99-ac27814b2af8.png)

Not pictured on this graph is the S3 throughput, due to an issue with the S3 sink not properly reporting metrics.  On a tweaked branch, running the `default_aws_es` variant, I verified that the S3 sink keeps pace without issue.

Once Bruce's ARC work is merged, and we can tweak the Kafka generator to produce more load, we'll do a final run of the `default_aws_es` and `concurrency_arc_aws_es` variants to make sure they still meet our OKR targets.
